src/feat/signal-distort.cc

```diff
@@ -72,5 +72,93 @@ void TimeStretch(const MatrixBase<BaseFloat> &input_egs,
 
+  // According to the rows of noise_eg, form the noise_mat
+  // repeat the noise_eg blocks to have a new block which is longer than input_eg
+ 
+  // As the noise_eg is very huge and the input_eg is small normally,
+  // so we'd better not reload the "noise_eg" matrix
+  // select the noise range
+
+  Matrix<BaseFloat> selected_noise_mat;
+  selected_noise_mat.Resize(input_rows, input_cols);
+  
+  int32 noise_rows = noise_eg.NumRows();
+  int32 start_row_ind = RandInt(0, noise_rows - input_rows);
+  
+  if (noise_eg.NumRows() < input_rows) {
+    int32 indices[input_rows];
+    for (int32 i=0; i < input_rows; ++i) {
+      indices[i] = (start_row_ind + i) % noise_eg.NumRows();
+    }
+    selected_noise_mat.CopyRows(noise_eg, indices);
+  } else {
+    selected_noise_mat.AddMat(1.0, noise_eg.Range(start_row_ind, input_rows,
+                                                  0, input_cols));
+  }
+
+  // compute the energy of noise and input
+  Matrix<BaseFloat> input_energy_mat(input_rows, input_cols);
+  input_energy_mat.AddMatMatElements(1.0, input_eg, input_eg, 0.0);
+  double input_energy = input_energy_mat.Sum();
+  Matrix<BaseFloat> noise_energy_mat(input_rows, input_cols);
+  noise_energy_mat.AddMatMatElements(1.0, selected_noise_mat, selected_noise_mat, 0.0);
+  double noise_energy = noise_energy_mat.Sum();
+
+  // In Energy domain, SNR=20log10(S/N). 
+  // 10^(SNR/20) = input_energy / (scale^2 * noise_energy)
+  double scale = input_energy / noise_energy / (pow(10,SNR/20));
+  scale = sqrt(scale);
+  
+  // Add noise mat to input_eg mat
+  perturb_eg->Resize(input_rows, input_cols);
+  perturb_eg->CopyFromMat(input_eg);
+  perturb_eg->AddMat(scale, selected_noise_mat);
+}
+
+void PerturbXvectorSignal::ApplyDistortion(const MatrixBase<BaseFloat> &input_egs,
+                                           Matrix<BaseFloat> *perturb_egs) {
+  if (!opts_.add_noise_rspecifier.empty()) { // deal with the add_noise ark situdation
+    // count the number of noise examples and record the key
+    std::vector<std::string> list_noise_egs;
+    list_noise_egs.clear();
+    kaldi::nnet3::SequentialNnetExampleReader noise_seq_reader(opts_.add_noise_rspecifier);
+    for (; !noise_seq_reader.Done(); noise_seq_reader.Next()) {
+      std::string key = noise_seq_reader.Key();
+      list_noise_egs.push_back(key);
+    }
+    noise_seq_reader.Close();
+    
+    // random choose a noise_eg and use it.
+    int32 num_noise_egs = list_noise_egs.size();
+    int32 index_noise_egs = RandInt(0, num_noise_egs - 1);
+    std::string key_noise_egs = list_noise_egs[index_noise_egs];
+
+    kaldi::nnet3::RandomAccessNnetExampleReader noise_random_reader(opts_.add_noise_rspecifier);
+    const kaldi::nnet3::NnetExample &noise_eg = noise_random_reader.Value(key_noise_egs);
+    const kaldi::nnet3::NnetIo &noise_eg_io = noise_eg.io[0];
+    Matrix<BaseFloat> noise_eg_mat;
+    noise_eg_io.features.CopyToMat(&noise_eg_mat);
+    int32 SNR = opts_.snr;
```

Comment:
Add these lines nnet3-perturb-egs binary.
I told you that You have to just call PerturbEgs function in nnet3-perturb-egs.cc. You should put loop for reading egs in nnet3 binary not here!
PerturbEgs should be defined as a separate function not a function of this class. The point is that in PertubEgs function, you create object from class PerturbXvectorSignal and call ApplyDistortion.